Targetd adds

tasleson opened the pull request:
Later versions of lvm removed the lvm2app API which targetd was
using.  Target has switched over to use libblockdev for its lvm
needs.  Thus targetd needs the ability to fork & exec lvm binary
which is what libblockdev does when using the lvm plugin.  Also
we need the ability to update the lvm hints file and talk to the lvm
dbus service.

Additional allows found by running new unit test.

see: https://github.com/fedora-selinux/selinux-policy/pull/431 for
added interfaces to lvm which this change is using.

zpytela commented on targetd.te:
```diff
@@ -25,13 +25,14 @@ files_tmp_file(targetd_tmp_t)
 
 
-allow targetd_t self:capability { ipc_lock sys_admin sys_nice };
+allow targetd_t self:capability { chown ipc_lock sys_admin sys_nice };
```
Do you know for which files this capability is required?

tasleson replied:
See https://github.com/fedora-selinux/selinux-policy-contrib/pull/334#issuecomment-694997933

zpytela commented on targetd.te:
```diff
@@ -48,6 +49,8 @@ fs_manage_configfs_files(targetd_t)
 fs_manage_configfs_lnk_files(targetd_t)
```
Why is access to unlabeled_t required? Which directory it is, which filesystem, do you have complete audit records?

tasleson replied:
See https://github.com/fedora-selinux/selinux-policy-contrib/pull/334#issuecomment-694997933

zpytela commented:
I'd also like to see the commit message be more descriptive. Just a bit, but it really helps.

tasleson commented:
I'll try to explain what targetd does so that maybe we can make the selinux
rules better.


targetd allows an API user to create a file system on the server and export it
over NFS.  Targetd has a configuration file `/etc/targetd/targetd.yaml` which
allows the user to define 1 or more btrfs file systems to use for the service.
When the API user creates a new filesystem we utilze btrfs subvolumes.  We also
allow the API user to create snapshots of the subvolumes among other things.

So an example:

Configuration file specifies a btrfs FS at `/mnt/btrfs`.  The targetd daemon
then creates the subvolumes targetd_fs and targetd_ss.  So the tree looks like

```
/mnt/btrfs/
├── targetd_fs
└── targetd_ss
```

Then when the API user creates a new FS is lands in `/mnt/btrfs/targetd_fs/*` and
snapsnots end up in `/mnt/btrfs/targetd_ss`.

We can also export a FS or FS SS over NFS and change the ownership of the mount
point through the API etc.

Ideally we would only need the permissions added to the `/mnt/btrfs/*` path,
but that path can be configured and changed by the user before daemon start.
There is probably a way for us to do this programmatically in the daemon
itself, which is probably the right answer, but I'm not a selinux guru.

Suggestions which don't require end users to do anything too complicated?

tasleson commented:
@zpytela Any updates?

zpytela commented:
@tasleson, I am sorry for the delay, but I needed to think about it for some time. Now I have a question: when lvm is executed, what type of access does it need since then?
Generally, when a different command is to be executed, we need to decide in the policy if to allow the execution in the caller's domain or a domain transition instead. I suppose most of the permission can be replaced with
```
lvm_domtrans(targetd_t)
```
but it really depends on which permissions are needed for targetd itself and which after the lvm execution.

For your path question: nothing to do in SELinux as long as context is correct.

I also don't understand when the unlabeled_t files come to play: does it happen to be the btrfs snapshot?

tasleson commented:
> @tasleson, I am sorry for the delay, but I needed to think about it for some time. Now I have a question: when lvm is executed, what type of access does it need since then?

I'm not exactly sure what you are asking here, but I think we only really need the lvm permissions to do lvm changes during the duration of the command execution.  After that targetd is doing a number of different things which are not lvm related.

> Generally, when a different command is to be executed, we need to decide in the policy if to allow the execution in the caller's domain or a domain transition instead. I suppose most of the permission can be replaced with
> 
> ```
> lvm_domtrans(targetd_t)
> ```
> 
> but it really depends on which permissions are needed for targetd itself and which after the lvm execution.
> 
> For your path question: nothing to do in SELinux as long as context is correct.
> 
> I also don't understand when the unlabeled_t files come to play: does it happen to be the btrfs snapshot?

If I take a block device, do mkfs.btrfs /dev/sdd and mount it at /mnt/btrfs I see this.

```
# ls -lZ
drwxr-xr-x. 1 root root system_u:object_r:unlabeled_t:s0 0 Sep 30 16:22 btrfs
```

If I do

```
# restorecon -v btrfs/
Relabeled /mnt/btrfs from system_u:object_r:unlabeled_t:s0 to system_u:object_r:mnt_t:s0
```
However, if I then create the subvolumes and list I get:

```
# btrfs subvolume create btrfs/targetd_fs
# btrfs subvolume create btrfs/targetd_ss

# ls -lZ
total 0
drwxr-xr-x. 1 root root system_u:object_r:unlabeled_t:s0 0 Sep 30 16:50 targetd_fs
drwxr-xr-x. 1 root root system_u:object_r:unlabeled_t:s0 0 Sep 30 16:50 targetd_ss
```
That is why I added unlabeled access as btrfs when it creates things, they are unlabeled.
This is also true if I create additional nested subvolumes or create a snapshot.

```
# btrfs subvolume snapshot -r targetd_fs/some_subvolume targetd_ss/some_sub_ss
Create a readonly snapshot of 'targetd_fs/some_subvolume' in 'targetd_ss/some_sub_ss'
# ls -lZ targetd_ss
total 0
drwxr-xr-x. 1 root root system_u:object_r:unlabeled_t:s0 0 Sep 30 17:03 some_sub_ss
```

zpytela commented:
> > @tasleson, I am sorry for the delay, but I needed to think about it for some time. Now I have a question: when lvm is executed, what type of access does it need since then?
> 
> I'm not exactly sure what you are asking here, but I think we only really need the lvm permissions to do lvm changes during the duration of the command execution. After that targetd is doing a number of different things which are not lvm related.

The options are just allow targetd execute lvm or allow transition to lvm's domain. It really looks like we could make targetd_t transition to lvm_t. Can you change the PR in this way and remove the rest of permissions/interfaces?

Then please share the AVC denials, e. g.
```
  # ausearch -i -m avc,user_avc,selinux_err,user_selinux_err -ts recent
```

> 
> > Generally, when a different command is to be executed, we need to decide in the policy if to allow the execution in the caller's domain or a domain transition instead. I suppose most of the permission can be replaced with
> > ```
> > lvm_domtrans(targetd_t)
> > ```
> > 
> > 
> > but it really depends on which permissions are needed for targetd itself and which after the lvm execution.
> > For your path question: nothing to do in SELinux as long as context is correct.
> > I also don't understand when the unlabeled_t files come to play: does it happen to be the btrfs snapshot?
> 
> If I take a block device, do mkfs.btrfs /dev/sdd and mount it at /mnt/btrfs I see this.
> 
> ```
> # ls -lZ
> drwxr-xr-x. 1 root root system_u:object_r:unlabeled_t:s0 0 Sep 30 16:22 btrfs
> ```
> 
> If I do
> 
> ```
> # restorecon -v btrfs/
> Relabeled /mnt/btrfs from system_u:object_r:unlabeled_t:s0 to system_u:object_r:mnt_t:s0
> ```
> 
> However, if I then create the subvolumes and list I get:
> 
> ```
> # btrfs subvolume create btrfs/targetd_fs
> # btrfs subvolume create btrfs/targetd_ss
> 
> # ls -lZ
> total 0
> drwxr-xr-x. 1 root root system_u:object_r:unlabeled_t:s0 0 Sep 30 16:50 targetd_fs
> drwxr-xr-x. 1 root root system_u:object_r:unlabeled_t:s0 0 Sep 30 16:50 targetd_ss
> ```
> 
> That is why I added unlabeled access as btrfs when it creates things, they are unlabeled.
> This is also true if I create additional nested subvolumes or create a snapshot.
> 
> ```
> # btrfs subvolume snapshot -r targetd_fs/some_subvolume targetd_ss/some_sub_ss
> Create a readonly snapshot of 'targetd_fs/some_subvolume' in 'targetd_ss/some_sub_ss'
> # ls -lZ targetd_ss
> total 0
> drwxr-xr-x. 1 root root system_u:object_r:unlabeled_t:s0 0 Sep 30 17:03 some_sub_ss
> ```

tasleson commented:
> The options are just allow targetd execute lvm or allow transition to lvm's domain. It really looks like we could make targetd_t transition to lvm_t. Can you change the PR in this way and remove the rest of permissions/interfaces?
> 
> Then please share the AVC denials, e. g.
> 
> ```
>   # ausearch -i -m avc,user_avc,selinux_err,user_selinux_err -ts recent
> ```

In progress, will let you know what I discover.

tasleson commented:
> The options are just allow targetd execute lvm or allow transition to lvm's domain. It really looks like we could make targetd_t transition to lvm_t. Can you change the PR in this way and remove the rest of permissions/interfaces?
> 
> Then please share the AVC denials, e. g.
> 
> ```
>   # ausearch -i -m avc,user_avc,selinux_err,user_selinux_err -ts recent
> ```

I made the following changes and I can run the unit test without errors or selinux denials.

``` patch
diff --git a/targetd.te b/targetd.te
index 4b760336b..32912449b 100644
--- a/targetd.te
+++ b/targetd.te
@@ -59,6 +59,9 @@ kernel_request_load_module(targetd_t)
 kernel_dgram_send(targetd_t)
 
 rpc_read_exports(targetd_t)
+rpc_write_exports(targetd_t)
+allow targetd_t exports_t:file { create unlink };
+allow targetd_t exports_t:dir { search read open write getattr add_name remove_name };
 
 storage_raw_rw_fixed_disk(targetd_t)
 
@@ -89,16 +92,7 @@ optional_policy(`
 ')
 
 optional_policy(`
-   lvm_read_config(targetd_t)
-   lvm_map_config(targetd_t)
-   lvm_write_metadata(targetd_t)
-   lvm_manage_metadata(targetd_t)
-   lvm_manage_lock(targetd_t)
-   lvm_rw_pipes(targetd_t)
-   lvm_stream_connect(targetd_t)
-   lvm_exec(targetd_t)
-   lvm_dbus_send_msg(targetd_t)
-   lvm_rw_var_run(targetd_t)
+    lvm_domtrans(targetd_t)
 ')
```
Changing lvm to ` lvm_domtrans(targetd_t)` works with no denials.  However, I ended up running into nfs export issues which are not related to this change.  I'm not sure how it worked before without them as targetd needs to be able to access files in `/etc/exports.d` and create, rename, delete etc.

Also I needed to do `chcon` on `/etc/exports.d` and `/etc/exports.d/targetd.exports`.  Otherwise they didn't have the correct context.  I'm thinking that should be added?

Let me know what you would like me to do.  I'm thinking we can back out the new interfaces in lvm, simplify targetd, and extend interface in rpc?  Also accommodate `/etc/exports.d` too?

tasleson committed:
Add interface rpc_manage_exports

Some applications have a need to manage nfs export files. Adding
interface to allow those applications full and complete access to
nfs export file life cycle.

Signed-off-by: Tony Asleson <tasleson@redhat.com>

tasleson committed:
Update targetd nfs & lvm

Simplify lvm by leveraging lvm_domtrans & adding the needed
capabilities to manage nfs export files.

Signed-off-by: Tony Asleson <tasleson@redhat.com>

tasleson commented:
@zpytela I've re-worked this PR.  I added some interfaces to `rpc` and changes lvm to `lvm_domtrans`.

tasleson commented:
These are the AVC denials for F32 with existing policy running in permissive mode.

```
# ausearch -i -m avc,user_avc,selinux_err,user_selinux_err -ts recent
----
type=AVC msg=audit(10/26/2020 13:51:33.746:182) : avc:  denied  { execute } for  pid=1253 comm=targetd name=lvm dev="sda3" ino=816308 scontext=system_u:system_r:targetd_t:s0 tcontext=system_u:object_r:lvm_exec_t:s0 tclass=file permissive=1 
----
type=AVC msg=audit(10/26/2020 13:51:33.747:183) : avc:  denied  { read open } for  pid=1253 comm=targetd path=/usr/sbin/lvm dev="sda3" ino=816308 scontext=system_u:system_r:targetd_t:s0 tcontext=system_u:object_r:lvm_exec_t:s0 tclass=file permissive=1 
----
type=AVC msg=audit(10/26/2020 13:51:33.747:184) : avc:  denied  { execute_no_trans } for  pid=1253 comm=targetd path=/usr/sbin/lvm dev="sda3" ino=816308 scontext=system_u:system_r:targetd_t:s0 tcontext=system_u:object_r:lvm_exec_t:s0 tclass=file permissive=1 
----
type=AVC msg=audit(10/26/2020 13:51:33.747:185) : avc:  denied  { map } for  pid=1253 comm=lvm path=/usr/sbin/lvm dev="sda3" ino=816308 scontext=system_u:system_r:targetd_t:s0 tcontext=system_u:object_r:lvm_exec_t:s0 tclass=file permissive=1 
----
type=AVC msg=audit(10/26/2020 13:51:33.759:186) : avc:  denied  { getattr } for  pid=1253 comm=lvm path=/run/lvm/hints dev="tmpfs" ino=21924 scontext=system_u:system_r:targetd_t:s0 tcontext=system_u:object_r:lvm_var_run_t:s0 tclass=file permissive=1 
----
type=AVC msg=audit(10/26/2020 13:51:33.759:187) : avc:  denied  { read write } for  pid=1253 comm=lvm name=hints dev="tmpfs" ino=21924 scontext=system_u:system_r:targetd_t:s0 tcontext=system_u:object_r:lvm_var_run_t:s0 tclass=file permissive=1 
----
type=AVC msg=audit(10/26/2020 13:51:33.759:188) : avc:  denied  { open } for  pid=1253 comm=lvm path=/run/lvm/hints dev="tmpfs" ino=21924 scontext=system_u:system_r:targetd_t:s0 tcontext=system_u:object_r:lvm_var_run_t:s0 tclass=file permissive=1 
----
type=AVC msg=audit(10/26/2020 13:51:33.759:189) : avc:  denied  { lock } for  pid=1253 comm=lvm path=/run/lvm/hints dev="tmpfs" ino=21924 scontext=system_u:system_r:targetd_t:s0 tcontext=system_u:object_r:lvm_var_run_t:s0 tclass=file permissive=1 
----
type=AVC msg=audit(10/26/2020 13:51:35.459:190) : avc:  denied  { setattr } for  pid=815 comm=targetd name=td_ifkpfonfjh dev="loop1" ino=256 scontext=system_u:system_r:targetd_t:s0 tcontext=system_u:object_r:unlabeled_t:s0 tclass=dir permissive=1 
----
type=AVC msg=audit(10/26/2020 13:51:35.459:191) : avc:  denied  { chown } for  pid=815 comm=targetd capability=chown  scontext=system_u:system_r:targetd_t:s0 tcontext=system_u:system_r:targetd_t:s0 tclass=capability permissive=1 
----
type=AVC msg=audit(10/26/2020 13:51:37.708:192) : avc:  denied  { read } for  pid=1546 comm=exportfs name=exports.d dev="sda3" ino=25571548 scontext=system_u:system_r:targetd_t:s0 tcontext=system_u:object_r:exports_t:s0 tclass=dir permissive=1 
----
type=AVC msg=audit(10/26/2020 13:51:37.708:193) : avc:  denied  { open } for  pid=1546 comm=exportfs path=/etc/exports.d dev="sda3" ino=25571548 scontext=system_u:system_r:targetd_t:s0 tcontext=system_u:object_r:exports_t:s0 tclass=dir permissive=1 
----
type=AVC msg=audit(10/26/2020 13:51:37.708:194) : avc:  denied  { getattr } for  pid=1546 comm=exportfs path=/etc/exports.d dev="sda3" ino=25571548 scontext=system_u:system_r:targetd_t:s0 tcontext=system_u:object_r:exports_t:s0 tclass=dir permissive=1 
----
type=AVC msg=audit(10/26/2020 13:51:37.708:195) : avc:  denied  { search } for  pid=1546 comm=exportfs name=exports.d dev="sda3" ino=25571548 scontext=system_u:system_r:targetd_t:s0 tcontext=system_u:object_r:exports_t:s0 tclass=dir permissive=1 
----
type=AVC msg=audit(10/26/2020 13:51:37.709:196) : avc:  denied  { write } for  pid=815 comm=targetd name=exports.d dev="sda3" ino=25571548 scontext=system_u:system_r:targetd_t:s0 tcontext=system_u:object_r:exports_t:s0 tclass=dir permissive=1 
----
type=AVC msg=audit(10/26/2020 13:51:37.709:197) : avc:  denied  { remove_name } for  pid=815 comm=targetd name=targetd.exports dev="sda3" ino=25569747 scontext=system_u:system_r:targetd_t:s0 tcontext=system_u:object_r:exports_t:s0 tclass=dir permissive=1 
----
type=AVC msg=audit(10/26/2020 13:51:37.709:198) : avc:  denied  { unlink } for  pid=815 comm=targetd name=targetd.exports dev="sda3" ino=25569747 scontext=system_u:system_r:targetd_t:s0 tcontext=system_u:object_r:exports_t:s0 tclass=file permissive=1 
----
type=AVC msg=audit(10/26/2020 13:51:37.710:199) : avc:  denied  { add_name } for  pid=815 comm=targetd name=targetd.exports scontext=system_u:system_r:targetd_t:s0 tcontext=system_u:object_r:exports_t:s0 tclass=dir permissive=1 
----
type=AVC msg=audit(10/26/2020 13:51:37.710:200) : avc:  denied  { create } for  pid=815 comm=targetd name=targetd.exports scontext=system_u:system_r:targetd_t:s0 tcontext=system_u:object_r:exports_t:s0 tclass=file permissive=1 
----
type=AVC msg=audit(10/26/2020 13:51:37.710:201) : avc:  denied  { write } for  pid=815 comm=targetd path=/etc/exports.d/targetd.exports dev="sda3" ino=25569747 scontext=system_u:system_r:targetd_t:s0 tcontext=system_u:object_r:exports_t:s0 tclass=file permissive=1 
```

tasleson commented:
Also my latest PR includes the `chown` and `setattr` because without them I get:

```
type=AVC msg=audit(10/26/2020 13:55:04.706:207) : avc:  denied  { setattr } for  pid=815 comm=targetd name=td_nstjiwxien dev="loop1" ino=256 scontext=system_u:system_r:targetd_t:s0 tcontext=system_u:object_r:unlabeled_t:s0 tclass=dir permissive=0 
----
type=AVC msg=audit(10/26/2020 13:55:07.154:208) : avc:  denied  { setattr } for  pid=815 comm=targetd name=td_uuarfxmmzq dev="loop1" ino=256 scontext=system_u:system_r:targetd_t:s0 tcontext=system_u:object_r:unlabeled_t:s0 tclass=dir permissive=0 
----
type=AVC msg=audit(10/26/2020 13:55:07.292:209) : avc:  denied  { setattr } for  pid=815 comm=targetd name=td_yreiywkgat dev="loop1" ino=256 scontext=system_u:system_r:targetd_t:s0 tcontext=system_u:object_r:unlabeled_t:s0 tclass=dir permissive=0 
----
type=AVC msg=audit(10/26/2020 14:05:42.198:214) : avc:  denied  { chown } for  pid=815 comm=targetd capability=chown  scontext=system_u:system_r:targetd_t:s0 tcontext=system_u:system_r:targetd_t:s0 tclass=capability permissive=0 
----
type=AVC msg=audit(10/26/2020 14:05:44.626:215) : avc:  denied  { chown } for  pid=815 comm=targetd capability=chown  scontext=system_u:system_r:targetd_t:s0 tcontext=system_u:system_r:targetd_t:s0 tclass=capability permissive=0 
----
type=AVC msg=audit(10/26/2020 14:05:44.765:216) : avc:  denied  { chown } for  pid=815 comm=targetd capability=chown  scontext=system_u:system_r:targetd_t:s0 tcontext=system_u:system_r:targetd_t:s0 tclass=capability permissive=0
```

tasleson commented:
@zpytela Any thoughts on latest?

zpytela commented:
@tasleson sorry for the delay. More information is needed. Does the service actually fail or is it just the denial you are concerned with?

The output misses information about syscalls; please remove the never audit rule if present, and add a void watch rule:
```
  # auditctl -l
-a never,task
  # auditctl -d never,task
  # auditctl -w /etc/shadow -p w -k shadow-write
```
Then run the scenario and collect denials in permissive mode, with the policy "after lvm_exec" and "before other changes".

Also note for accessing types from a different module you need to use an interface, or create one if it does not exist.

tasleson commented:
> @tasleson sorry for the delay. More information is needed. Does the service actually fail or is it just the denial you are concerned with?

It actually fails.  I'll supply you with the additional information you request tomorrow, thanks!

tasleson commented:
@zpytela  I ran the unit test with the following changes and the system in permissive mode.  Please note that in actual user case running, the service will be making changes against block devices that aren't loop back devices.

```bash
# auditctl -l
-w /etc/shadow -p w -k shadow-write
```

```diff
diff --git a/targetd.te b/targetd.te
index b7f5164d3..64ee5c590 100644
--- a/targetd.te
+++ b/targetd.te
@@ -86,13 +86,7 @@ optional_policy(`
 ')
 
 optional_policy(`
-   lvm_read_config(targetd_t)
-   lvm_map_config(targetd_t)
-   lvm_write_metadata(targetd_t)
-   lvm_manage_metadata(targetd_t)
-   lvm_manage_lock(targetd_t)
-   lvm_rw_pipes(targetd_t)
-   lvm_stream_connect(targetd_t)
+    lvm_domtrans(targetd_t)
 ')
 
 optional_policy(`
```

```
[root@localhost targetd-devel]# ausearch -i -m avc,user_avc,selinux_err,user_selinux_err -ts recent
----
type=PROCTITLE msg=audit(11/20/2020 10:43:19.536:218) : proctitle=targetd 
type=PATH msg=audit(11/20/2020 10:43:19.536:218) : item=0 name=/mnt/btrfs/targetd_fs/td_hgutccsnkw inode=256 dev=00:27 mode=dir,755 ouid=root ogid=root rdev=00:00 obj=system_u:object_r:unlabeled_t:s0 nametype=NORMAL cap_fp=none cap_fi=none cap_fe=0 cap_fver=0 cap_frootid=0 
type=CWD msg=audit(11/20/2020 10:43:19.536:218) : cwd=/ 
type=SYSCALL msg=audit(11/20/2020 10:43:19.536:218) : arch=x86_64 syscall=chown success=yes exit=0 a0=0x7f6c93461d20 a1=unknown(1000) a2=unset a3=0x7ffe1ffe7a37 items=1 ppid=1 pid=867 auid=unset uid=root gid=root euid=root suid=root fsuid=root egid=root sgid=root fsgid=root tty=(none) ses=unset comm=targetd exe=/usr/bin/python3.8 subj=system_u:system_r:targetd_t:s0 key=(null) 
type=AVC msg=audit(11/20/2020 10:43:19.536:218) : avc:  denied  { chown } for  pid=867 comm=targetd capability=chown  scontext=system_u:system_r:targetd_t:s0 tcontext=system_u:system_r:targetd_t:s0 tclass=capability permissive=1 
type=AVC msg=audit(11/20/2020 10:43:19.536:218) : avc:  denied  { setattr } for  pid=867 comm=targetd name=td_hgutccsnkw dev="loop1" ino=256 scontext=system_u:system_r:targetd_t:s0 tcontext=system_u:object_r:unlabeled_t:s0 tclass=dir permissive=1 
----
type=PROCTITLE msg=audit(11/20/2020 10:43:21.804:219) : proctitle=exportfs -o insecure 0.0.0.0/0:/mnt/btrfs/targetd_fs/td_dhnegmhgmj 
type=PATH msg=audit(11/20/2020 10:43:21.804:219) : item=0 name=/etc/exports.d inode=25571548 dev=08:03 mode=dir,755 ouid=root ogid=root rdev=00:00 obj=system_u:object_r:exports_t:s0 nametype=NORMAL cap_fp=none cap_fi=none cap_fe=0 cap_fver=0 cap_frootid=0 
type=CWD msg=audit(11/20/2020 10:43:21.804:219) : cwd=/ 
type=SYSCALL msg=audit(11/20/2020 10:43:21.804:219) : arch=x86_64 syscall=openat success=yes exit=4 a0=0xffffff9c a1=0x55a59095e19b a2=O_RDONLY|O_NONBLOCK|O_DIRECTORY|O_CLOEXEC a3=0x0 items=1 ppid=867 pid=1211 auid=unset uid=root gid=root euid=root suid=root fsuid=root egid=root sgid=root fsgid=root tty=(none) ses=unset comm=exportfs exe=/usr/sbin/exportfs subj=system_u:system_r:targetd_t:s0 key=(null) 
type=AVC msg=audit(11/20/2020 10:43:21.804:219) : avc:  denied  { open } for  pid=1211 comm=exportfs path=/etc/exports.d dev="sda3" ino=25571548 scontext=system_u:system_r:targetd_t:s0 tcontext=system_u:object_r:exports_t:s0 tclass=dir permissive=1 
type=AVC msg=audit(11/20/2020 10:43:21.804:219) : avc:  denied  { read } for  pid=1211 comm=exportfs name=exports.d dev="sda3" ino=25571548 scontext=system_u:system_r:targetd_t:s0 tcontext=system_u:object_r:exports_t:s0 tclass=dir permissive=1 
----
type=PROCTITLE msg=audit(11/20/2020 10:43:21.804:220) : proctitle=exportfs -o insecure 0.0.0.0/0:/mnt/btrfs/targetd_fs/td_dhnegmhgmj 
type=SYSCALL msg=audit(11/20/2020 10:43:21.804:220) : arch=x86_64 syscall=fstat success=yes exit=0 a0=0x4 a1=0x7ffcc5172cf0 a2=0x7ffcc5172cf0 a3=0x0 items=0 ppid=867 pid=1211 auid=unset uid=root gid=root euid=root suid=root fsuid=root egid=root sgid=root fsgid=root tty=(none) ses=unset comm=exportfs exe=/usr/sbin/exportfs subj=system_u:system_r:targetd_t:s0 key=(null) 
type=AVC msg=audit(11/20/2020 10:43:21.804:220) : avc:  denied  { getattr } for  pid=1211 comm=exportfs path=/etc/exports.d dev="sda3" ino=25571548 scontext=system_u:system_r:targetd_t:s0 tcontext=system_u:object_r:exports_t:s0 tclass=dir permissive=1 
----
type=PROCTITLE msg=audit(11/20/2020 10:43:21.805:221) : proctitle=exportfs -o insecure 0.0.0.0/0:/mnt/btrfs/targetd_fs/td_dhnegmhgmj 
type=PATH msg=audit(11/20/2020 10:43:21.805:221) : item=0 name=/etc/exports.d/targetd.exports inode=25569728 dev=08:03 mode=file,644 ouid=root ogid=root rdev=00:00 obj=system_u:object_r:exports_t:s0 nametype=NORMAL cap_fp=none cap_fi=none cap_fe=0 cap_fver=0 cap_frootid=0 
type=CWD msg=audit(11/20/2020 10:43:21.805:221) : cwd=/ 
type=SYSCALL msg=audit(11/20/2020 10:43:21.805:221) : arch=x86_64 syscall=openat success=yes exit=4 a0=0xffffff9c a1=0x7ffcc5172df0 a2=O_RDONLY a3=0x0 items=1 ppid=867 pid=1211 auid=unset uid=root gid=root euid=root suid=root fsuid=root egid=root sgid=root fsgid=root tty=(none) ses=unset comm=exportfs exe=/usr/sbin/exportfs subj=system_u:system_r:targetd_t:s0 key=(null) 
type=AVC msg=audit(11/20/2020 10:43:21.805:221) : avc:  denied  { search } for  pid=1211 comm=exportfs name=exports.d dev="sda3" ino=25571548 scontext=system_u:system_r:targetd_t:s0 tcontext=system_u:object_r:exports_t:s0 tclass=dir permissive=1 
----
type=PROCTITLE msg=audit(11/20/2020 10:43:21.807:222) : proctitle=targetd 
type=PATH msg=audit(11/20/2020 10:43:21.807:222) : item=1 name=/etc/exports.d/targetd.exports inode=25569728 dev=08:03 mode=file,644 ouid=root ogid=root rdev=00:00 obj=system_u:object_r:exports_t:s0 nametype=DELETE cap_fp=none cap_fi=none cap_fe=0 cap_fver=0 cap_frootid=0 
type=PATH msg=audit(11/20/2020 10:43:21.807:222) : item=0 name=/etc/exports.d/ inode=25571548 dev=08:03 mode=dir,755 ouid=root ogid=root rdev=00:00 obj=system_u:object_r:exports_t:s0 nametype=PARENT cap_fp=none cap_fi=none cap_fe=0 cap_fver=0 cap_frootid=0 
type=CWD msg=audit(11/20/2020 10:43:21.807:222) : cwd=/ 
type=SYSCALL msg=audit(11/20/2020 10:43:21.807:222) : arch=x86_64 syscall=unlink success=yes exit=0 a0=0x7f6c935cce90 a1=0x0 a2=0x0 a3=0x7ffe1ffe7797 items=2 ppid=1 pid=867 auid=unset uid=root gid=root euid=root suid=root fsuid=root egid=root sgid=root fsgid=root tty=(none) ses=unset comm=targetd exe=/usr/bin/python3.8 subj=system_u:system_r:targetd_t:s0 key=(null) 
type=AVC msg=audit(11/20/2020 10:43:21.807:222) : avc:  denied  { unlink } for  pid=867 comm=targetd name=targetd.exports dev="sda3" ino=25569728 scontext=system_u:system_r:targetd_t:s0 tcontext=system_u:object_r:exports_t:s0 tclass=file permissive=1 
type=AVC msg=audit(11/20/2020 10:43:21.807:222) : avc:  denied  { remove_name } for  pid=867 comm=targetd name=targetd.exports dev="sda3" ino=25569728 scontext=system_u:system_r:targetd_t:s0 tcontext=system_u:object_r:exports_t:s0 tclass=dir permissive=1 
type=AVC msg=audit(11/20/2020 10:43:21.807:222) : avc:  denied  { write } for  pid=867 comm=targetd name=exports.d dev="sda3" ino=25571548 scontext=system_u:system_r:targetd_t:s0 tcontext=system_u:object_r:exports_t:s0 tclass=dir permissive=1 
----
type=PROCTITLE msg=audit(11/20/2020 10:43:21.808:223) : proctitle=targetd 
type=PATH msg=audit(11/20/2020 10:43:21.808:223) : item=1 name=/etc/exports.d/targetd.exports inode=25569728 dev=08:03 mode=file,644 ouid=root ogid=root rdev=00:00 obj=system_u:object_r:exports_t:s0 nametype=CREATE cap_fp=none cap_fi=none cap_fe=0 cap_fver=0 cap_frootid=0 
type=PATH msg=audit(11/20/2020 10:43:21.808:223) : item=0 name=/etc/exports.d/ inode=25571548 dev=08:03 mode=dir,755 ouid=root ogid=root rdev=00:00 obj=system_u:object_r:exports_t:s0 nametype=PARENT cap_fp=none cap_fi=none cap_fe=0 cap_fver=0 cap_frootid=0 
type=CWD msg=audit(11/20/2020 10:43:21.808:223) : cwd=/ 
type=SYSCALL msg=audit(11/20/2020 10:43:21.808:223) : arch=x86_64 syscall=openat success=yes exit=7 a0=0xffffff9c a1=0x7f6c935d4e50 a2=O_WRONLY|O_CREAT|O_TRUNC|O_CLOEXEC a3=0x1b6 items=2 ppid=1 pid=867 auid=unset uid=root gid=root euid=root suid=root fsuid=root egid=root sgid=root fsgid=root tty=(none) ses=unset comm=targetd exe=/usr/bin/python3.8 subj=system_u:system_r:targetd_t:s0 key=(null) 
type=AVC msg=audit(11/20/2020 10:43:21.808:223) : avc:  denied  { write } for  pid=867 comm=targetd path=/etc/exports.d/targetd.exports dev="sda3" ino=25569728 scontext=system_u:system_r:targetd_t:s0 tcontext=system_u:object_r:exports_t:s0 tclass=file permissive=1 
type=AVC msg=audit(11/20/2020 10:43:21.808:223) : avc:  denied  { create } for  pid=867 comm=targetd name=targetd.exports scontext=system_u:system_r:targetd_t:s0 tcontext=system_u:object_r:exports_t:s0 tclass=file permissive=1 
type=AVC msg=audit(11/20/2020 10:43:21.808:223) : avc:  denied  { add_name } for  pid=867 comm=targetd name=targetd.exports scontext=system_u:system_r:targetd_t:s0 tcontext=system_u:object_r:exports_t:s0 tclass=dir permissive=1 
----
type=PROCTITLE msg=audit(11/20/2020 10:43:21.968:224) : proctitle=targetd 
type=PATH msg=audit(11/20/2020 10:43:21.968:224) : item=0 name=/mnt/btrfs/targetd_fs/td_jznttqlwqj inode=256 dev=00:27 mode=dir,755 ouid=root ogid=root rdev=00:00 obj=system_u:object_r:unlabeled_t:s0 nametype=NORMAL cap_fp=none cap_fi=none cap_fe=0 cap_fver=0 cap_frootid=0 
type=CWD msg=audit(11/20/2020 10:43:21.968:224) : cwd=/ 
type=SYSCALL msg=audit(11/20/2020 10:43:21.968:224) : arch=x86_64 syscall=chown success=yes exit=0 a0=0x7f6c93485c80 a1=unknown(1000) a2=unset a3=0x7ffe1ffe7a37 items=1 ppid=1 pid=867 auid=unset uid=root gid=root euid=root suid=root fsuid=root egid=root sgid=root fsgid=root tty=(none) ses=unset comm=targetd exe=/usr/bin/python3.8 subj=system_u:system_r:targetd_t:s0 key=(null) 
type=AVC msg=audit(11/20/2020 10:43:21.968:224) : avc:  denied  { setattr } for  pid=867 comm=targetd name=td_jznttqlwqj dev="loop1" ino=256 scontext=system_u:system_r:targetd_t:s0 tcontext=system_u:object_r:unlabeled_t:s0 tclass=dir permissive=1 
```

zpytela commented on targetd.te:
```diff
@@ -23,13 +23,14 @@ files_tmp_file(targetd_tmp_t)
 allow targetd_t self:tcp_socket { accept listen };
 allow targetd_t self:netlink_route_socket r_netlink_socket_perms;
 allow targetd_t self:process { setfscreate setsched };
+allow targetd_t unlabeled_t:dir setattr;
```
Please use `files_setattr_isid_type_dirs` interface.

tasleson replied:
Done

zpytela commented on rpc.if:
```diff
@@ -132,6 +132,43 @@ interface(`rpc_write_exports',`
+		type exports_t;
+	')
+
+	allow $1 exports_t:dir { search read open write getattr add_name remove_name };
```
Instead of enumerating permissions, use `rw_dir_perms` pattern, or even `manage_files_pattern()` interface which could also cover the files handling.

tasleson replied:
Removed the two interfaces I created in rpc and replaced with a single `rpc_manage_exports` which uses the `manage_files_pattern`

zpytela commented on rpc.if:
```diff
@@ -132,6 +132,43 @@ interface(`rpc_write_exports',`
+
+########################################
+## <summary>
+##	Allow export directory operations
```
The summary can be more descriptive: which operations?

tasleson replied:
Interface removed

zpytela commented on rpc.if:
```diff
@@ -132,6 +132,43 @@ interface(`rpc_write_exports',`
+##	</summary>
+## </param>
+#
+interface(`rpc_add_rm_exports',`
```
Is it necessary or useful to discern between files and directories managing?
Interfaces names tend to use read/write/rw/manage/etc. in their second part.

tasleson replied:
Interface removed

zpytela commented on targetd.te:
```diff
@@ -56,6 +57,8 @@ kernel_request_load_module(targetd_t)
 
 rpc_read_exports(targetd_t)
+rpc_write_exports(targetd_t)
+rpc_manage_exports(targetd_t)
```
If I am not mistaken, _manage_ is a complete superset of _write_ here.

tasleson replied:
I actually forgot to amend latest changes which removed the `rpc_write_exports`.  Just refreshed.

zpytela commented:
The policy LGTM now. Can you confirm it working?

tasleson commented:
> The policy LGTM now. Can you confirm it working?

Policy is working with the unit test case

zpytela commented:
@tasleson, thank you for this PR.

tasleson commented:
@zpytela Would it be possible to get this picked up in F33 & F32?

zpytela commented:
@tasleson here we go:
https://github.com/fedora-selinux/selinux-policy-contrib/pull/378
https://github.com/fedora-selinux/selinux-policy-contrib/pull/379

tasleson commented:
@zpytela Thanks
